src/curl.c

```diff
@@ -539,7 +526,91 @@ static int swupd_curl_get_file_full(const char *url, char *filename,
 		unlink(filename);
 	}
 
-	return err;
+	return -status;
```

Comment:
You could return status here and change the return of this function from in to enum download_status

Reply:
The problem here is that the **_process_curl_error_codes_** function which is called by **_swupd_curl_get_file_full_** used to return a negative errno number, and with this refactoring those errnos were removed from the function so it now returns a positive **_enum download_status_**. Most functions that make use of this function only check for values !=0 to determine if the file was gotten successfully by curl, but some functions (like _get_latest_version_ rely on the negative value that the function used to return). So bottom line is that to make this change we will need to scan many other functions to make sure they won't break if we change the return value from negative to positive.

Reply:
The problem with the suggested approach is that it is inconsistent. It's not a good idea to use different return codes in  the same function. For example, in this case, if curl hasn't been initialized, or if there's any error in curl api calls, swupd_curl_get_file_full returns -1. When you use this value by doing determine_strategy(-ret), we assume the status code is 1 (DOWNLOAD_STATUS_PARTIAL_FILE), what's incorrect.

You don't need to propagate the return value all over. The only user of swupd_curl_get_file_full is retry_download, so it's ok to change the return code from download_status.

And I would change retry_download to download a negative errno code, to be compatible with other functions in the API.

Reply:
Done